src/qat/model/device.py

```diff
@@ -0,0 +1,117 @@
+    block_size: Optional[int] = Field(ge=1, default=1)
+    phase_iq_offset: float = 0.0
+    bias: float = 1.0
+    acquire_allowed: bool = False
+
+
+class PulseChannel(Component):
+    """
+    Models a pulse channel on a particular device.
+    Attributes:
+        physical_channel: Physical channel that carries the pulse.
+        frequency: Frequency of the pulse.
+        bias: Mean value of the signal, quantifies offset relative to zero mean.
+        scale: Scale factor for mapping the voltage of the pulse to frequencies.
+        fixed_if: Flag which determines if the intermediate frequency is fixed.
+        channel_type: Type of the pulse.
+        auxiliary_qubits: Any extra devices this PulseChannel could be affecting except
+                           the current one. For example in cross resonance pulses.
+    """
+
+    physical_channel: Ref[PhysicalChannel]
+
+    frequency: float = Field(ge=0.0, default=0.0)
+    bias: complex = 0.0 + 0.0j
+    scale: complex = 1.0 + 0.0j
+    fixed_if: bool = False
+
+    channel_type: Optional[ChannelType] = Field(frozen=True, default=None)
+    auxiliary_qubits: RefList[Qubit] = []
+
+
+class QuantumDevice(Component):
+    """
+    A physical device whose main form of operation involves pulse channels.
+    Attributes:
+        pulse_channels: Pulse channels with their ids as keys.
+        physical_channel: Physical channel associated with the pulse channels.
+                          Note that this physical channel must be equal to the
+                          physical channel associated with the pulse channels.
+        measure_device: A quantum device used to measure the state of this device.
+        default_pulse_channel_type: Default type of pulse for the quantum device.
+    """
+
+    pulse_channels: RefDict[PulseChannel] = Field(frozen=True)
+    physical_channel: Ref[PhysicalChannel] = Field(frozen=True)
+    measure_device: Ref[Resonator] = Field(frozen=True)
+
+    default_pulse_channel_type: ChannelType = Field(
+        frozen=True, default=ChannelType.measure
+    )
+
+
+class Resonator(QuantumDevice):
+    """Models a resonator on a chip. Can be connected to multiple qubits."""
+
+    measure_device: None = Field(frozen=True, default=None)
+
+
+class Qubit(QuantumDevice):
+    """
+    Models a superconducting qubit on a chip, and holds all information relating to it.
+    Attributes:
+        measure_device: The resonator coupled to the qubit.
+        index: The index of the qubit on the chip.
+        drive_amp: Amplitude for the pulse controlling the |0> -> |1> transition of the qubit.
+        default_pulse_channel_type: Default type of pulse for the qubit.
+    """
+
+    measure_device: Ref[Resonator] = Field(frozen=True)
+
+    index: int = Field(ge=0)
+    drive_amp: float = 1.0
+    default_pulse_channel_type: ChannelType = Field(frozen=True, default=ChannelType.drive)
+
+    coupled_qubits: Optional[RefList[Qubit]] = None
```

Comment:
Do we have any plans for how to deal with the pulse property dicts in legacy Qubits? i.e. `pulse_hw_x_pi_2`,  `pulse_measure` and `measure_acquire `

Reply:
Yes that will be implemented in the next stage.